csrc/dynamic_transform.cpp

```diff
@@ -385,6 +648,219 @@ void DynamicTransformConcretizer::concretize() {
+          std::vector<TensorView*> nonempty_inputs;
+          for (auto inp : cop->inputs()) {
+            // Each "input" to CatOp is a pad() of the corresponding _actual_
+            // input. Here we peel off the pad op to collect the non-padded cat
+            // inputs.
+            auto padded_inp_tv = inp->as<TensorView>();
+            TORCH_INTERNAL_ASSERT(
+                padded_inp_tv->definition() &&
+                    padded_inp_tv->definition()->isA<PadOp>(),
+                "Input to cat should have definition that is a PadOp");
+            auto inp_tv = padded_inp_tv->definition()
+                              ->as<PadOp>()
+                              ->in()
+                              ->as<TensorView>();
+
+            if (inp_tv != tv) {
+              // we could remove other empty tensors here while we're at it.
+              // They will get removed by further passes anyway though as tv
+              // ranges over all empty tensors.
+              nonempty_inputs.push_back(inp_tv);
+            }
+          }
+          auto old_cat = cop->output(0)->as<TensorView>();
+          auto new_cat = nonempty_inputs.size() == 1
+              ? set(nonempty_inputs[0])
+              : cat(nonempty_inputs, cop->concatenatedDim());
+          replaceTV(old_cat, new_cat);
+        } else { // Replace pads that are not part of CatOps with full()
+          auto out_shape = orig_shape(out);
+          // Wherever there is a zero in the input, we will replace the original
+          // output extent so that we no longer reference the now-zero input
+          // extent
+          for (auto i : empty_tv_descr.empty_axes) {
+            auto pad_widths = pop->getPadWidths((int)i);
+            out_shape[i] = add(pad_widths.first, pad_widths.second);
+          }
+          replaceWithFull(out, out_shape, pop->value());
+        }
+      }
+    }
+    if (!tv->isFusionInput()) {
+      replaceWithFull(tv, new_shape);
+    }
+  }
+}
+
+TensorView* DynamicTransformConcretizer::replaceWithFull(
+    TensorView* tv,
+    std::vector<Val*>& new_shape,
+    Val* fill_value) {
+  TensorView* mut_tv = nullptr;
+  if (!tv->definition()) {
+    return tv;
+  }
+  if (!fill_value) {
+    fill_value = tv->fusion()->zeroVal();
+  }
+  if (fill_value->getDataType().value() != tv->getDataType().value()) {
+    fill_value = castOp(tv->getDataType().value(), fill_value);
+  }
+  mut_tv = full(new_shape, fill_value, tv->getDataType().value());
+
+  replaceTV(tv, mut_tv);
+
+  return mut_tv;
+}
+
+void DynamicTransformConcretizer::replaceTV(
+    TensorView* old_tv,
+    TensorView* new_tv) {
+  registerConcretization(old_tv, new_tv);
+  OptOutMutator::mutate(old_tv);
+
+  for (auto use : old_tv->uses()) {
+    ir_utils::replaceValInExpr(use, old_tv, new_tv);
+  }
+
+  if (old_tv->isFusionOutput()) {
+    old_tv->fusion()->replaceOutput(old_tv, new_tv);
+  }
+
+  replaced_tvs_[old_tv] = new_tv;
```

Comment:
Sorry, what do you mean by "only mutates and does not replace it"? Is this a generic issue with the mutator interface? If so, does it make sense to have a generic fix?

Reply:
In concretization, we have so far not wanted to replace `TensorView`s, but with this PR (and #511) we have started rewriting the `TensorView` graph instead of just updating the `TensorDomain`s, like we do for `resize` and `reshape`. 

Instead of changing `OptOutMutator`, I think I just need to fix how we are using it. I had a misunderstanding about how it was meant to be used and thought that if we did not want to directly replace an expression, that registering the input or output `Val`s would be enough. This is not true since `OptOutMutator::mutate(Expr*)` is actually the thing that does val replacement. So one needs to traverse the graph in topological order and call `mutate(stmt)` for each `Expr` as well as any `Val`s that might need to be changed (including `TensorView`s and `TensorDomain`s if they contain `IterDomain`s that have been changed). Note that replacing inputs or outputs would need to be done manually and can be done at any point. I think currently we are not calling `mutate` on exprs during the concretization loop, but if we do that (and also call `replaceOutput` when needed) then marking vals for replacement should work fine.

I will try this and report back.

Reply:
>In concretization, we have so far not wanted to replace TensorViews, but with this PR (and https://github.com/NVIDIA/Fuser/pull/511) we have started rewriting the TensorView graph instead of just updating the TensorDomains, like we do for resize and reshape.

Right, that's true.

Reply:
I've almost converted the code but it is fragile and it is starting to feel like this doesn't actually belong in concretization. Really, this is an optimization that we could do as an optimization pass if we set enough information at concretization. The reason I put it here originally was that we need the change to be reflected in concretization info since it depends on input sizes. But altering the `TensorView` graph is really what the optimization passes are for.

Instead, we could do the following at concretization without ever replacing one `TensorView` with another:

- Concretize `reshape` as we do now: by replacing the rfactor domain with the transforms found in `analyzeView`.
- Set constant extents of `zeroVal()` wherever we find empty tensors (replaces this PR).
- Concretize `slice` by setting the output extents according to the concrete branch (replaces #511). We would skip replacing the output with `full` (for empty slices) and `set` (trivial slices), but we would set the output extents to either `zeroVal()` or the input extent in those cases so that we can detect them in an optimization pass.
- Propagate concretized extents and `IterType`s as we do now.

Then, we would add optimization passes to do the following rewriting of the graph:

- Detect empty tensors by looking for constant zero extents in a backward traversal pass. Replace them with `full` if they are outputs, otherwise replace their reduction/pad/cat uses as appropriate (replaces this PR).
- Detect size-1 extents in outputs from operations like pad or slice. Currently we set the output `IterType` to broadcast at concretization, but we could additionally translate them into more meaningful and possibly simpler operations. For example, if we pad with non-positive pad widths and the result has size 1, it is actually a `select`+`broadcast`. In order to determine if the pad widths are non-positive, we could use axioms (#440). We could additionally check that the broadcast is resolved here since that may help simplify the graph as well (see #340).
- Detect trivial `slice` outputs by checking whether their slice `start`, `stop`, `step` are `sameAs` 0, `input->extent()`, and 1; if so we would replace the `slice` with `set`.

Reply:
I have opened #543 which addresses this much more cleanly I think. In this new approach we will only modify `TensorDomain`s during concretization, and we will leave rewriting the `TensorView`s for the pre-segmentation passes. I am now working on simplifying the present PR so that we just set extents to zero in the concretization pass.